packages/members-api/lib/stripe/index.js

```diff
@@ -125,73 +133,114 @@ module.exports = class StripePaymentProcessor {
+        await this._updateCustomer(member, customer);
+        if (!customer.subscriptions || !customer.subscriptions.data) {
             return;
         }
+        for (const subscription of customer.subscriptions.data) {
+            await this._updateSubscription(subscription);
+        }
+    }
+
+    async handleCustomerSubscriptionDeletedWebhook(subscription) {
+        await this._updateSubscription(subscription);
+    }
+
+    async handleCustomerSubscriptionUpdatedWebhook(subscription) {
+        await this._updateSubscription(subscription);
+    }
+
+    async handleInvoicePaymentSucceededWebhook(invoice) {
+        await this._updateSubscription(invoice.subscription);
+    }
+
+    async handleInvoicePaymentFailedWebhook(invoice) {
+        await this._updateSubscription(invoice.subscription);
+    }
+
+    async _updateCustomer(member, customer) {
         debug(`Attaching customer to member ${member.email} ${customer.id}`);
-        return this.storage.set(member, metadata.concat({
-            customer_id: customer.id
-        }));
+        await this.storage.set({
+            customer: {
+                customer_id: customer.id,
+                member_id: member.id,
+                name: customer.name,
+                email: customer.email
+            }
+        });
+    }
+
+    async _updateSubscription(subscription) {
+        const payment = subscription.default_payment_method;
+        if (typeof payment === 'string') {
+            debug(`Fetching default_payment_method for subscription ${subscription.id}`);
+            const subscriptionWithPayment = await retrieve(this._stripe, 'subscriptions', subscription.id, {
+                expand: ['default_payment_method']
+            });
+            return this._updateSubscription(subscriptionWithPayment);
+        }
+        debug(`Attaching subscription to customer ${subscription.customer} ${subscription.id}`);
+        await this.storage.set({
+            subscription: {
+                customer_id: subscription.customer,
+
+                subscription_id: subscription.id,
+                status: subscription.status,
+                current_period_end: new Date(subscription.current_period_end * 1000),
+                start_date: new Date(subscription.start_date * 1000),
+                default_payment_card_last4: payment && payment.card && payment.card.last4 || null,
+
+                plan_id: subscription.plan.id,
+                plan_nickname: subscription.plan.nickname,
+                plan_interval: subscription.plan.interval,
+                plan_amount: subscription.plan.amount,
+                plan_currency: subscription.plan.currency
+            }
+        });
     }
 
-    async _customerForMember(member) {
+    async _customerForMemberCheckoutSession(member) {
         const metadata = await this.storage.get(member);
 
-        for (const data in metadata) {
+        for (const data in metadata.customers) {
```

Comment:
@allouis  Was stuck debugging a problem where I had multiple customers created when trying to update payment info for a single customer. Tracked it down to this change. The issue here is that the `metadata.customers` is always an array, and when iterated with `for..in` it only looks up through keys which are 0,1,2 and so on for arrays. Thus this logic always jumps over and creates a new customer.
This logic doesn't seem intentional, but still wanted to double-check as we might have been creating some extra customers for members :grimacing:  

Think the correct way to handle legic here would be through regular array iteration like done [above](https://github.com/TryGhost/Members/pull/90/files#diff-10fda4eee1ef37a24be4fc806994d2a0R136). So would be something like `metadata.customers.forEach...` Let me know!

Reply:
Ah yeah good catch - this should be `for (const data of...`